### PR TITLE
Separate IOMMU and PCI device checks.

### PR DESCRIPTION
**Pull request by ebeaty-cisco**
### Summary
`host-check` checks that there are PCI devices available in the IOMMU check, but IOMMU is only relevant for `vfio-pci`, not `igb_uio`. Should separate the checks for PCI devices from the IOMMU check.

New check setup:
* `check_iommu()`: keeps existing checks for IOMMU but no longer looks for PCI devices.
* `check_pci_devices()`:
    *  If IOMMU is enabled, checks that IOMMU PCI devices are available.
    *  If IOMMU is not enabled, checks that any PCI devices are available.

Testing: existing tests for IOMMU split between new IOMMU and PCI devices checks. 

### Checklist

- [ ] Changelog updated (or not required)
  
- [ ] Static analysis and tests passing
   - Use `commit-check` script, or check GitHub Actions status.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -5,6 +5,10 @@
 
 The v1 release supports Cisco IOS-XR release versions from 7.7.1 to 7.10.1.
 
+### v1.1.14 (2024-03-04)
```
Comment (tjohnes) [inline CHANGELOG.md:8]:
```suggestion
### v1.1.15 (2024-03-04)
```

File: scripts/host-check [hunk 2]
```diff
@@ -1126,31 +1139,32 @@ def check_iommu() -> CheckFuncReturn:
...
-        )
-    net_iommu_devs_list = []
-    for device in sorted(net_iommu_devices):
-        net_iommu_devs_list.append(
-            net_devices_dict[device] + " (" + device + ")"
+            "determine the network devices on the host.",
         )
+
+    # If in IOMMU mode then check if the network devices are in the IOMMU group
+    if iommu_devices:
```
Comment (tjohnes) [inline scripts/host-check:1146]: `iommu_devices` may not be defined I think? (if we fall into the `except` on L1119)

File: scripts/host-check [hunk 3]
```diff
@@ -1126,31 +1139,32 @@ def check_iommu() -> CheckFuncReturn:
...
+        network_devices &= set(iommu_devices)
+        if not network_devices:
+            return (
+                CheckState.WARNING,
+                "IOMMU enabled for vfio-pci, but no network PCI devices found.",
```
Comment (tjohnes) [inline scripts/host-check:1151]: Should this just be "No network PCI devices found." now? (or "no PCI network devices found" like the above warning)

File: scripts/host-check (hunk 2 above)
Comment (tjohnes) [reply scripts/host-check:1146]: (if so there should be a test to cover this)

Comment (ebeaty-cisco) [reply scripts/host-check:1146]: Changed the except so that iommu_devices = [] if the directory search fails.

File: scripts/host-check [hunk 4]
```diff
@@ -1112,8 +1127,6 @@ def check_iommu() -> CheckFuncReturn:
             r"pci@([\da-f]{4}:[\da-f]{2}:[\da-f]{2}\.[\da-f])\s+(\S+)", output
         )
         if not matches:
-            # This should always be a warning - we need PCI devices whether
```
[the remark this answers is not in the document]
Comment (ebeaty-cisco) [reply scripts/host-check:1115]: The comment refers to the fact that for igb_uio, most errors in check_iommu were neutral except this one. As this is now in check_pci_devices, all errors are treated the same regardless of driver and so the comment isn't relevant

File: tests/test_host_check.py [hunk 5]
```diff
@@ -2992,107 +2991,11 @@ def test_success(self, capsys):
...
-            PASS -- IOMMU
-                    IOMMU enabled for vfio-pci with the following PCI device(s):
-                    device1 (0000:00:00.0), device2 (0000:00:01.0), device3 (0000:00:02.0),
-                    device4 (0000:00:1f.2)
+            PASS -- IOMMU (IOMMU enabled for vfio-pci.)
             """
         )
         assert result is CheckState.SUCCESS
 
-    def test_no_iommu_devices(self, capsys):
```
Comment (tjohnes) [inline tests/test_host_check.py:3003]: This test case is still relevant for the IOMMU check, I think - if there are no IOMMU devices (i.e., `iommu_devices = []`), the check should emit a warning (NEUTRAL if uio present, else WARNING).

File: tests/test_host_check.py [hunk 6]
```diff
@@ -3267,6 +3167,167 @@ def test_warn_igb_uio_enabled(self, capsys):
...
+class TestPCIDevices(_CheckTestBase):
+    """Tests for the PCI devices check."""
+
+    check_group = "xrd-vrouter"
+    check_name = "PCI devices"
+    deps = ["IOMMU"]
+    cmds = ["lshw -businfo -c network"]
+    files = []
+
+    def test_success_vfio_iommu(self, capsys):
```
Comment (tjohnes) [inline tests/test_host_check.py:3179]: I don't see how these tests stub/mock which drivers are available? (vfio/uio)

File: tests/test_host_check.py [hunk 7]
```diff
@@ -2992,107 +2991,11 @@ def test_success(self, capsys):
...
-                    IOMMU enabled for vfio-pci with the following PCI device(s):
-                    device1 (0000:00:00.0), device2 (0000:00:01.0), device3 (0000:00:02.0),
-                    device4 (0000:00:1f.2)
+            PASS -- IOMMU (IOMMU enabled for vfio-pci.)
             """
         )
         assert result is CheckState.SUCCESS
 
-    def test_no_iommu_devices(self, capsys):
-        """Test the case where no devices are found."""
```
Comment (tjohnes) [inline tests/test_host_check.py:3004]: I don't see how these tests stub/mock what drivers are available either - i.e., there should be several cases here:

- vfio not present => NEUTRAL
- If vfio present then define "warning" as NEUTRAL if uio is present else WARNING in the following test cases:
  - `enable_unsafe_noiommu_mode` is Y => warning
  - For (`enable_unsafe_noiommu_mode` is N, `enable_unsafe_noiommu_mode` is not present):
    - `/sys/class/iommu/*/devices/*` glob is empty => warning
    - There's an error case that's probably not worth worrying about (not sure how `glob.glob` raises an Exception?
    - Else success.

Comment (tjohnes) [reply tests/test_host_check.py:3004]: Actually I see now that what drivers are available is controlled by `cmd_effects`.  This is really quite hard to follow, but I think this is:

```
                cmd_effects=[
                    "",  # vfio-pci is supported
                    None,  # skip?????
                    subprocess.SubprocessError,  # uio is not supported?
                    subprocess.SubprocessError,  # not sure what this ever was
                    lshw_output,  # not relevant anymore
                ],

File: tests/test_host_check.py (hunk 6 above)
Comment (tjohnes) [reply tests/test_host_check.py:3179]: In fact, `check_pci_devices` no longer cares what drivers are available, so not sure why there are test cases distinguishing between the drivers?

File: tests/test_host_check.py [hunk 8]
```diff
@@ -3267,6 +3167,167 @@ def test_warn_igb_uio_enabled(self, capsys):
...
+====================================================
+pci@0000:00:00.0  device1     network    82540EM Gigabit Ethernet Controller
+pci@0000:00:1f.2  device4     network    Ethernet interface
+pci@0000:00:02.0  device3     network    82540EM Gigabit Ethernet Controller
+pci@0000:00:01.0  device2     network    Ethernet interface
+                  docker0     network    Ethernet interface
+
+
+        """
+        result, output = self.perform_check(
```
Comment (tjohnes) [inline tests/test_host_check.py:3227]: I'm not sure how this works without mocking `glob.glob`?!

File: scripts/host-check (hunk 2 above)
Comment (tjohnes) [reply scripts/host-check:1147]: Resolving this - I'm not sure what the correct behaviour should be here if there are no IOMMU devices, but moving this to a different thread.

File: scripts/host-check [hunk 9]
```diff
@@ -1126,31 +1140,32 @@ def check_iommu() -> CheckFuncReturn:
...
+            "determine the network devices on the host.",
         )
+
+    # If in IOMMU mode then check if the network devices are in the IOMMU group
+    if iommu_devices:
+        network_devices &= set(iommu_devices)
+        if not network_devices:
+            return (
+                CheckState.WARNING,
+                "No PCI network devices found with IOMMU enabled for vfio-pci.",
```
Comment (tjohnes) [inline scripts/host-check:1152]: How do we know we are using vfio-pci?

File: scripts/host-check [hunk 10]
```diff
@@ -1126,31 +1140,32 @@ def check_iommu() -> CheckFuncReturn:
...
-    for device in sorted(net_iommu_devices):
-        net_iommu_devs_list.append(
-            net_devices_dict[device] + " (" + device + ")"
+            "determine the network devices on the host.",
         )
+
+    # If in IOMMU mode then check if the network devices are in the IOMMU group
+    if iommu_devices:
+        network_devices &= set(iommu_devices)
+        if not network_devices:
```
Comment (tjohnes) [inline scripts/host-check:1149]: Per the PR description:

If IOMMU is enabled, checks that IOMMU PCI devices are available.
If IOMMU is not enabled, checks that any PCI devices are available.

but here, if IOMMU is not enabled (which we are saying is equivalent to `iommu_devices` is empty?), this check returns WARNING?

File: scripts/host-check [hunk 11]
```diff
@@ -1104,6 +1104,22 @@ def check_iommu() -> CheckFuncReturn:
             f"Unable to check if IOMMU is enabled by listing {iommu_dev_filepath}.\n"
             + recommendation,
         )
+    return (CheckState.SUCCESS, "IOMMU enabled for vfio-pci.")
+
+
+def check_pci_devices() -> CheckFuncReturn:
```
Comment (tjohnes) [inline scripts/host-check:1110]: Per the PR description:

If IOMMU is enabled, checks that IOMMU PCI devices are available.
If IOMMU is not enabled, checks that any PCI devices are available.

I'm not convinced this is the desirable behaviour.  Instead should this be:

- If uio is present: check any PCI devices are available and list them all (customer can use uio and not worry about IOMMU; this is the same as `check_iommu`, which returns NEUTRAL if uio and vfio are present but IOMMU is disabled)
- If only vfio is present: check IOMMU PCI devices are available and list them, if not return WARNING.

File: scripts/host-check (hunk 10 above)
Comment (ebeaty-cisco) [reply scripts/host-check:1149]: This return CheckState.WARNING (on L1150) is hit if both
* `iommu_devices` is not empty list - indicates IOMMU is enabled (see L1147)
* There are no network devices in the IOMMU group.

File: scripts/host-check (hunk 11 above)
Comment (ebeaty-cisco) [reply scripts/host-check:1110]: Yeah, I think that sounds right.

File: scripts/host-check (hunk 10 above)
Comment (ebeaty-cisco) [reply scripts/host-check:1149]: (but subject to change, cf. comment at L1110)

File: scripts/host-check (hunk 11 above)
Comment (ebeaty-cisco) [reply scripts/host-check:1110]: Added a new require_iommu flag which is set to true if vfio is supported and uio is not, otherwise false. Then the check only requires IOMMU PCI devices if this flag is true.

File: scripts/host-check (hunk 9 above)
Comment (ebeaty-cisco) [reply scripts/host-check:1152]: With other changes to check_pci_devices, this is now indicated by require_iommu flag

File: tests/test_host_check.py (hunk 5 above)
Comment (ebeaty-cisco) [reply tests/test_host_check.py:3003]: I don't think so, this test is actually for the case that no iommu pci devices are present, ie `iommu_devices & network_devices = []`, which is no longer part of `check_iommu`

File: tests/test_host_check.py (hunk 7 above)
Comment (ebeaty-cisco) [reply tests/test_host_check.py:3004]: for each check if a driver is supported, there can be up to two commands. The second is only called if the first fails so for vfio-pci in this case:
```
"", # => vfio-pci is supported
None, # this command is skipped
```
and for uio
```
subprocess.SubprocessError, # the first command returns an error
subprocess.SubprocessError, # so the second command is run, as both fail uio is not supported

File: tests/test_host_check.py (hunk 6 above)
Comment (ebeaty-cisco) [reply tests/test_host_check.py:3179]: Test cases cover both the case that IOMMU is required (i.e. uio not supported and no-iommu not set) and not

File: tests/test_host_check.py (hunk 8 above)
Comment (ebeaty-cisco) [reply tests/test_host_check.py:3227]: Yeah, `glob.glob` should be mocked here (I guess it worked anyway due to the exception catching)

File: scripts/host-check (hunk 11 above)
Comment (tjohnes) [reply scripts/host-check:1110]: > except for the case that vfio is explicitly in non-IOMMU mode

This sounds reasonable - except there's now a bit of an inconsistency in that if we are using vfio only, and there are no `iommu_devices` but `enable_unsafe_noiommu_mode` is N or not present, then `check_iommu` is WARNING, and `check_pci_devices` is also WARNING ("No PCI network devices found with IOMMU enabled for vfio-pci").

Either this case should also list all PCI devices, or we should just list IOMMU PCI devices always for vfio only (I don't have a strong preference but lean slightly towards the latter).

The other option which maybe makes more sense from an end-user point of view would be to SKIP `check_pci_devices` if vfio only and `check_iommu` was WARN - but this is too much of a design change (SKIP is treated as a failure)

Comment (tjohnes) [reply scripts/host-check:1110]: (see `test_no_iommu_devices` :))

File: tests/test_host_check.py (hunk 5 above)
Comment (tjohnes) [reply tests/test_host_check.py:3003]: Is there another test case covering "no IOMMU devices"?

File: tests/test_host_check.py (hunk 7 above)
Comment (tjohnes) [reply tests/test_host_check.py:3004]: Ah gotcha

File: tests/test_host_check.py [hunk 12]
```diff
@@ -3267,6 +3167,206 @@ def test_warn_igb_uio_enabled(self, capsys):
...
+                    The following PCI device(s) are available:
+                    device1 (0000:00:00.0), device2 (0000:00:01.0), device3 (0000:00:02.0),
+                    device4 (0000:00:1f.2)
+            """
+        )
+        assert result is CheckState.SUCCESS
+
+    def test_no_iommu_devices(self, capsys):
+        """
+        Test the case where no IOMMU devices are found when in IOMMU mode.
```
Comment (tjohnes) [inline tests/test_host_check.py:3266]:
```suggestion
        Test the case where no IOMMU network devices are found when in IOMMU mode.
```

File: scripts/host-check (hunk 11 above)
Comment (tjohnes) [reply scripts/host-check:1110]: Summary of offline discussion:
- If `/sys/module/vfio/parameters/enable_unsafe_noiommu_mode` is Y, then we assume the user knows what they are doing and its more useful for them to see all PCI devices.
- In contrast, there may be some IOMMU devices (so `check_iommu` passes), but no IOMMU netdevices.  In this case we shouldn't list the non-IOMMU netdevices.

Keep the behaviour as proposed - list IOMMU netdevices, unless the `/sys/module/vfio/parameters/enable_unsafe_noiommu_mode` is Y, in which case the user knows what they are doing (and they are already warned against doing this in `check_iommu`), so give them all netdevices.

**Review by tjohnes — APPROVED** [review]
